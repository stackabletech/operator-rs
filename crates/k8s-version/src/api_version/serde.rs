use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::ApiVersion;

impl<'de> Deserialize<'de> for ApiVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct ApiVersionVisitor;

        impl Visitor<'_> for ApiVersionVisitor {
            type Value = ApiVersion;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                ApiVersion::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(ApiVersionVisitor)
    }
}

impl Serialize for ApiVersion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

#[cfg(feature = "serde")]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize() {
        let _: ApiVersion =
            serde_yaml::from_str("extensions.k8s.io/v1alpha1").expect("api version is valid");
    }

    #[test]
    fn serialize() {
        let api_version =
            ApiVersion::from_str("extensions.k8s.io/v1alpha1").expect("api version is valid");
        assert_eq!(
            "extensions.k8s.io/v1alpha1\n",
            serde_yaml::to_string(&api_version).expect("api version must serialize")
        );
    }
}
