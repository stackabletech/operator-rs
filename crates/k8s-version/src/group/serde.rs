use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Group;

impl<'de> Deserialize<'de> for Group {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct GroupVisitor;

        impl<'de> Visitor<'de> for GroupVisitor {
            type Value = Group;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API group")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Group::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(GroupVisitor)
    }
}

impl Serialize for Group {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self)
    }
}
