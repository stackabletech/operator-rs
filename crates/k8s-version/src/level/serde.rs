use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Level;

impl<'de> Deserialize<'de> for Level {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct LevelVisitor;

        impl<'de> Visitor<'de> for LevelVisitor {
            type Value = Level;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version level")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Level::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(LevelVisitor)
    }
}

impl Serialize for Level {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

#[cfg(feature = "serde")]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize() {
        let _: Level = serde_yaml::from_str("alpha1").expect("level is valid");
    }

    #[test]
    fn serialize() {
        let api_version = Level::from_str("alpha1").expect("level is valid");
        assert_eq!(
            "alpha1\n",
            serde_yaml::to_string(&api_version).expect("level must serialize")
        );
    }
}
