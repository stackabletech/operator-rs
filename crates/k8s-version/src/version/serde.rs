use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Version;

impl<'de> Deserialize<'de> for Version {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct VersionVisitor;

        impl<'de> Visitor<'de> for VersionVisitor {
            type Value = Version;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Version::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(VersionVisitor)
    }
}

impl Serialize for Version {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}
