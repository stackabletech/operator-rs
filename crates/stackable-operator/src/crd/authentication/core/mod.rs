use kube::CustomResource;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use stackable_versioned::versioned;

mod v1alpha1_impl;

#[versioned(version(name = "v1alpha1"))]
pub mod versioned {
    // This makes v1alpha1 versions of all authentication providers available to the
    // AuthenticationClassProvider enum below.
    mod v1alpha1 {
        use crate::crd::authentication::{kerberos, ldap, oidc, r#static, tls};
    }
    /// The Stackable Platform uses the AuthenticationClass as a central mechanism to handle user
    /// authentication across supported products.
    ///
    /// The authentication mechanism needs to be configured only in the AuthenticationClass which is
    /// then referenced in the product. Multiple different authentication providers are supported.
    /// Learn more in the [authentication concept documentation][1] and the
    /// [Authentication with OpenLDAP tutorial][2].
    ///
    /// [1]: DOCS_BASE_URL_PLACEHOLDER/concepts/authentication
    /// [2]: DOCS_BASE_URL_PLACEHOLDER/tutorials/authentication_with_openldap
    #[versioned(k8s(
        group = "authentication.stackable.tech",
        plural = "authenticationclasses",
        crates(
            kube_core = "kube::core",
            k8s_openapi = "k8s_openapi",
            schemars = "schemars"
        )
    ))]
    #[derive(
        Clone,
        Debug,
        Eq,
        Hash,
        Ord,
        PartialEq,
        PartialOrd,
        CustomResource,
        Deserialize,
        JsonSchema,
        Serialize,
    )]
    #[serde(rename_all = "camelCase")]
    pub struct AuthenticationClassSpec {
        /// Provider used for authentication like LDAP or Kerberos.
        pub provider: AuthenticationClassProvider,
    }

    #[derive(
        Clone,
        Debug,
        Deserialize,
        strum::Display,
        Eq,
        Hash,
        JsonSchema,
        Ord,
        PartialEq,
        PartialOrd,
        Serialize,
    )]
    #[serde(rename_all = "camelCase")]
    #[allow(clippy::large_enum_variant)]
    pub enum AuthenticationClassProvider {
        /// The [static provider](https://DOCS_BASE_URL_PLACEHOLDER/concepts/authentication#_static)
        /// is used to configure a static set of users, identified by username and password.
        Static(r#static::v1alpha1::AuthenticationProvider),

        /// The [LDAP provider](DOCS_BASE_URL_PLACEHOLDER/concepts/authentication#_ldap).
        /// There is also the ["Authentication with LDAP" tutorial](DOCS_BASE_URL_PLACEHOLDER/tutorials/authentication_with_openldap)
        /// where you can learn to configure Superset and Trino with OpenLDAP.
        Ldap(ldap::v1alpha1::AuthenticationProvider),

        /// The OIDC provider can be used to configure OpenID Connect.
        Oidc(oidc::v1alpha1::AuthenticationProvider),

        /// The [TLS provider](DOCS_BASE_URL_PLACEHOLDER/concepts/authentication#_tls).
        /// The TLS AuthenticationClass is used when users should authenticate themselves with a TLS certificate.
        Tls(tls::v1alpha1::AuthenticationProvider),

        /// The [Kerberos provider](DOCS_BASE_URL_PLACEHOLDER/concepts/authentication#_kerberos).
        /// The Kerberos AuthenticationClass is used when users should authenticate themselves via Kerberos.
        Kerberos(kerberos::v1alpha1::AuthenticationProvider),
    }

    /// Common [`v1alpha1::ClientAuthenticationDetails`] which is specified at the client/ product
    /// cluster level. It provides a name (key) to resolve a particular [`AuthenticationClass`].
    /// Additionally, it provides authentication provider specific configuration (OIDC and LDAP for
    /// example).
    ///
    /// If the product needs additional (product specific) authentication options, it is recommended
    /// to wrap this struct and use `#[serde(flatten)]` on the field.
    ///
    /// Additionally, it might be the case that special fields are needed in the contained structs,
    /// such as [`oidc::v1alpha1::ClientAuthenticationOptions`]. To be able to add custom fields in
    /// that structs without serde(flattening) multiple structs, they are generic, so you can add
    /// additional attributes if needed.
    ///
    /// ### Example
    ///
    /// ```
    /// # use schemars::JsonSchema;
    /// # use serde::{Deserialize, Serialize};
    /// use stackable_operator::crd::authentication::core::v1alpha1;
    ///
    /// #[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
    /// #[serde(rename_all = "camelCase")]
    /// pub struct SupersetAuthenticationClass {
    ///     pub user_registration_role: String,
    ///     pub user_registration: bool,
    ///
    ///     #[serde(flatten)]
    ///     pub common: v1alpha1::ClientAuthenticationDetails,
    /// }
    /// ```
    #[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
    #[serde(rename_all = "camelCase")]
    #[schemars(description = "")]
    pub struct ClientAuthenticationDetails<O = ()> {
        /// Name of the [AuthenticationClass](https://docs.stackable.tech/home/nightly/concepts/authentication) used to
        /// authenticate users.
        //
        // To get the concrete [`AuthenticationClass`], we must resolve it. This resolution can be achieved by using
        // [`ClientAuthenticationDetails::resolve_class`].
        #[serde(rename = "authenticationClass")]
        authentication_class_ref: String,

        /// This field contains OIDC-specific configuration. It is only required in case OIDC is used.
        //
        // Use [`ClientAuthenticationDetails::oidc_or_error`] to get the value or report an error to the user.
        // TODO: Ideally we want this to be an enum once other `ClientAuthenticationOptions` are added, so
        // that user can not configure multiple options at the same time (yes we are aware that this makes a
        // changing the type of an AuthenticationClass harder).
        // This is a non-breaking change though :)
        oidc: Option<oidc::v1alpha1::ClientAuthenticationOptions<O>>,
    }
}
