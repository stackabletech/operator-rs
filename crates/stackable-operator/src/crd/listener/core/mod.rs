use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use stackable_versioned::versioned;

#[cfg(doc)]
use crate::crd::listener::class::v1alpha1::ListenerClass;

mod v1alpha1_impl;

#[versioned(version(name = "v1alpha1"))]
pub mod versioned {
    /// The method used to access the services.
    //
    // Please note that this does not necessarily need to be restricted to the same Service types
    // Kubernetes supports. Listeners currently happens to support the same set of service types as
    // upstream Kubernetes, but we still want to have the freedom to add custom ones in the future
    // (for example: Istio ingress?).
    #[derive(Serialize, Deserialize, Clone, Copy, Debug, JsonSchema, PartialEq, Eq)]
    pub enum ServiceType {
        /// Reserve a port on each node.
        NodePort,

        /// Provision a dedicated load balancer.
        LoadBalancer,

        /// Assigns an IP address from a pool of IP addresses that your cluster has reserved for
        /// that purpose.
        ClusterIP,
    }

    /// Service Internal Traffic Policy enables internal traffic restrictions to only route internal
    /// traffic to endpoints within the node the traffic originated from. The "internal" traffic
    /// here refers to traffic originated from Pods in the current cluster. This can help to reduce
    /// costs and improve performance. See [Kubernetes docs][k8s-docs].
    ///
    /// [k8s-docs]: https://kubernetes.io/docs/concepts/services-networking/service-traffic-policy/
    //
    // Please note that this represents a Kubernetes type, so the name of the enum variant needs to
    // exactly match the Kubernetes traffic policy.
    #[derive(Serialize, Deserialize, Clone, Debug, JsonSchema, PartialEq, Eq, strum::Display)]
    pub enum KubernetesTrafficPolicy {
        /// Obscures the client source IP and may cause a second hop to another node, but allows
        /// Kubernetes to spread the load between all nodes.
        Cluster,

        /// Preserves the client source IP and avoid a second hop for LoadBalancer and NodePort type
        /// Services, but makes clients responsible for spreading the load.
        ///
        /// Does not work everywhere, e.g. not on IONOS!
        Local,
    }

    /// The type of a given address.
    #[derive(Serialize, Deserialize, Clone, Copy, Debug, JsonSchema, PartialEq, Eq)]
    #[serde(rename_all = "PascalCase")]
    pub enum AddressType {
        /// A resolvable DNS hostname.
        Hostname,

        /// A resolved IP address.
        #[serde(rename = "IP")]
        Ip,
    }

    /// A mode for deciding the preferred [`v1alpha1::AddressType`].
    ///
    /// These can vary depending on the rest of the [`ListenerClass`][lc].
    ///
    /// [lc]: crate::crd::listener::class::v1alpha1::ListenerClass
    #[derive(Serialize, Deserialize, Clone, Copy, Debug, JsonSchema, PartialEq, Eq)]
    pub enum PreferredAddressType {
        /// Like [`AddressType::Hostname`], but prefers [`AddressType::Ip`] for [`ServiceType::NodePort`], since their hostnames are less likely to be resolvable.
        HostnameConservative,

        // Like the respective variants of AddressType. Ideally we would refer to them instead of
        // copy/pasting, but that breaks due to upstream issues:
        //
        // - https://github.com/GREsau/schemars/issues/222
        // - https://github.com/kube-rs/kube/issues/1622
        Hostname,

        #[serde(rename = "IP")]
        Ip,
    }
}
