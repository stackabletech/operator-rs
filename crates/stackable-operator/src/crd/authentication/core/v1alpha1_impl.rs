use snafu::{OptionExt as _, Snafu};

use crate::{
    client::Client,
    crd::authentication::{
        core::v1alpha1::{AuthenticationClass, ClientAuthenticationDetails},
        oidc::v1alpha1 as oidc_v1alpha1,
    },
};

type Result<T, E = Error> = std::result::Result<T, E>;

// NOTE (@Techassi): Where is the best place to put this?
#[derive(Debug, PartialEq, Snafu)]
pub enum Error {
    #[snafu(display("authentication details for OIDC were not specified. The AuthenticationClass {auth_class_name:?} uses an OIDC provider, you need to specify OIDC authentication details (such as client credentials) as well"))]
    OidcAuthenticationDetailsNotSpecified { auth_class_name: String },
}

impl AuthenticationClass {
    pub async fn resolve(
        client: &Client,
        authentication_class_name: &str,
    ) -> crate::client::Result<AuthenticationClass> {
        client
            .get::<AuthenticationClass>(authentication_class_name, &()) // AuthenticationClass has ClusterScope
            .await
    }
}

impl<O> ClientAuthenticationDetails<O> {
    /// Resolves this specific [`AuthenticationClass`]. Usually products support
    /// a list of authentication classes, which individually need to be resolved.crate::client
    pub async fn resolve_class(
        &self,
        client: &Client,
    ) -> crate::client::Result<AuthenticationClass> {
        AuthenticationClass::resolve(client, &self.authentication_class_ref).await
    }

    pub fn authentication_class_name(&self) -> &String {
        &self.authentication_class_ref
    }

    /// In case OIDC is configured, the user *needs* to provide some connection details,
    /// such as the client credentials. Call this function in case the user has configured
    /// OIDC, as it will error out then the OIDC client details are missing.
    pub fn oidc_or_error(
        &self,
        auth_class_name: &str,
    ) -> Result<&oidc_v1alpha1::ClientAuthenticationOptions<O>> {
        self.oidc
            .as_ref()
            .with_context(|| OidcAuthenticationDetailsNotSpecifiedSnafu {
                auth_class_name: auth_class_name.to_string(),
            })
    }
}

#[cfg(test)]
mod tests {
    use crate::crd::authentication::{
        core::v1alpha1::AuthenticationClassProvider, kerberos::v1alpha1 as kerberos_v1alpha1,
        tls::v1alpha1 as tls_v1alpha1,
    };

    #[test]
    fn provider_to_string() {
        let tls_provider = AuthenticationClassProvider::Tls(tls_v1alpha1::AuthenticationProvider {
            client_cert_secret_class: None,
        });
        assert_eq!("Tls", tls_provider.to_string());

        let kerberos_provider =
            AuthenticationClassProvider::Kerberos(kerberos_v1alpha1::AuthenticationProvider {
                kerberos_secret_class: "kerberos".to_string(),
            });
        assert_eq!("Kerberos", kerberos_provider.to_string());
    }
}
