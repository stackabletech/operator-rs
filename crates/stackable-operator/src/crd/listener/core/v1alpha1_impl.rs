use crate::crd::listener::{
    class::v1alpha1::ListenerClassSpec,
    core::v1alpha1::{AddressType, PreferredAddressType, ServiceType},
};

impl PreferredAddressType {
    pub fn resolve(self, listener_class: &ListenerClassSpec) -> AddressType {
        match self {
            PreferredAddressType::HostnameConservative => match listener_class.service_type {
                ServiceType::NodePort => AddressType::Ip,
                _ => AddressType::Hostname,
            },
            PreferredAddressType::Hostname => AddressType::Hostname,
            PreferredAddressType::Ip => AddressType::Ip,
        }
    }
}
