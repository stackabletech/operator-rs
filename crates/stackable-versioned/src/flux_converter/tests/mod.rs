use kube::CustomResource;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use stackable_versioned_macros::versioned;

use crate as stackable_versioned;

#[versioned(
    k8s(group = "test.stackable.tech",),
    version(name = "v1alpha1"),
    version(name = "v1alpha2"),
    version(name = "v1beta1"),
    version(name = "v2"),
    version(name = "v3")
)]
#[derive(
    Clone,
    Debug,
    Eq,
    Hash,
    Ord,
    PartialEq,
    PartialOrd,
    CustomResource,
    Deserialize,
    JsonSchema,
    Serialize,
)]
#[serde(rename_all = "camelCase")]
struct PersonSpec {
    username: String,

    // In v1alpha2 first and last name have been added
    #[versioned(added(since = "v1alpha2"))]
    first_name: String,
    #[versioned(added(since = "v1alpha2"))]
    last_name: String,

    // We started out with a enum. As we *need* to provide a default, we have a Unknown variant.
    // Afterwards we figured let's be more flexible and accept any arbitrary String.
    #[versioned(
        added(since = "v2", default = "default_gender"),
        changed(since = "v3", from_type = "Gender")
    )]
    gender: String,
}

fn default_gender() -> Gender {
    Gender::Unknown
}

#[derive(
    Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Deserialize, JsonSchema, Serialize,
)]
#[serde(rename_all = "PascalCase")]
pub enum Gender {
    Unknown,
    Male,
    Female,
}

impl Into<String> for Gender {
    fn into(self) -> String {
        match self {
            Gender::Unknown => "Unknown".to_string(),
            Gender::Male => "Male".to_string(),
            Gender::Female => "Female".to_string(),
        }
    }
}
impl From<String> for Gender {
    fn from(value: String) -> Self {
        match value.as_str() {
            "Male" => Self::Male,
            "Female" => Self::Female,
            _ => Self::Unknown,
        }
    }
}

/// TEMP, we need to implement downgrades manually
impl From<v1alpha2::PersonSpec> for v1alpha1::PersonSpec {
    fn from(value: v1alpha2::PersonSpec) -> Self {
        Self {
            username: value.username,
        }
    }
}
impl From<v1beta1::PersonSpec> for v1alpha2::PersonSpec {
    fn from(value: v1beta1::PersonSpec) -> Self {
        Self {
            username: value.username,
            first_name: value.first_name,
            last_name: value.last_name,
        }
    }
}
impl From<v2::PersonSpec> for v1beta1::PersonSpec {
    fn from(value: v2::PersonSpec) -> Self {
        Self {
            username: value.username,
            first_name: value.first_name,
            last_name: value.last_name,
        }
    }
}
impl From<v3::PersonSpec> for v2::PersonSpec {
    fn from(value: v3::PersonSpec) -> Self {
        Self {
            username: value.username,
            first_name: value.first_name,
            last_name: value.last_name,
            gender: value.gender.into(),
        }
    }
}
/// END TEMP

#[cfg(test)]
mod tests {
    use std::{fs::File, path::Path};

    use insta::{assert_snapshot, glob};
    use kube::core::{
        conversion::{ConversionResponse, ConversionReview},
        response::StatusSummary,
    };

    use super::Person;

    #[test]
    fn pass() {
        glob!("../../../fixtures/inputs/pass/", "*.json", |path| {
            let (review, response) = run_for_file(path);

            assert_eq!(response.result.status, Some(StatusSummary::Success));
            assert_eq!(review.request.unwrap().uid, response.uid);

            let formatted = serde_json::to_string_pretty(&response)
                .expect("Failed to serialize ConversionResponse");
            assert_snapshot!(formatted);
        })
    }

    #[test]
    fn fail() {
        glob!("../../../fixtures/inputs/fail/", "*.json", |path| {
            let (review, response) = run_for_file(path);

            assert_eq!(response.result.status, Some(StatusSummary::Failure));
            if let Some(request) = &review.request {
                assert_eq!(request.uid, response.uid);
            }

            let formatted = serde_json::to_string_pretty(&response)
                .expect("Failed to serialize ConversionResponse");
            assert_snapshot!(formatted);
        })
    }

    fn run_for_file(path: &Path) -> (ConversionReview, ConversionResponse) {
        let review: ConversionReview =
            serde_json::from_reader(File::open(path).expect("failed to open test file"))
                .expect("failed to parse ConversionReview from test file");
        let response = Person::convert(review.clone());

        (review, response)
    }
}
