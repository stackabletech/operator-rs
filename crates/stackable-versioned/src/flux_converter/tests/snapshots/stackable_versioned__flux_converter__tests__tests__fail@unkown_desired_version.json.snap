---
source: crates/stackable-versioned/src/flux_converter/tests/mod.rs
expression: formatted
input_file: crates/stackable-versioned/fixtures/inputs/fail/unkown_desired_version.json
---
{
  "apiVersion": "apiextensions.k8s.io/v1",
  "kind": "ConversionReview",
  "response": {
    "uid": "c4e55572-ee1f-4e94-9097-28936985d45f",
    "result": {
      "status": "Failure",
      "code": 500,
      "message": "failed to parse desired resource version \"test.stackable.tech/v99\": the api version \"test.stackable.tech/v99\" is not known",
      "reason": "failed to parse desired resource version \"test.stackable.tech/v99\": the api version \"test.stackable.tech/v99\" is not known"
    },
    "convertedObjects": []
  }
}
